File: native/bindings/endstone/level/block.cpp
#include "block.h"
#include "../../../aegilex_types.h"

#include <aegilex-runtime/src/cxx_host_level.rs.h>

#include <endstone/block/block.h>
#include <endstone/block/block_data.h>
#include <endstone/block/block_state.h>
#include <endstone/level/dimension.h>
#include <endstone/level/location.h>
#include <endstone/server.h>

#include <cstdint>
#include <memory>
#include <string>
#include <string_view>
#include <type_traits>
#include <variant>

namespace aegilex::native::level {

class Block::impl {
  public:
    explicit impl(std::unique_ptr<endstone::Block> block, endstone::Server *server) noexcept
        : block(std::move(block)), server(server)
    {
    }

    std::unique_ptr<endstone::Block> block;
    endstone::Server *server;
};


namespace {

constexpr std::uint32_t kBlockStateBoolean = 0U;
constexpr std::uint32_t kBlockStateText = 1U;
constexpr std::uint32_t kBlockStateInteger = 2U;

[[nodiscard]] aegilex::status block_states_from_pairs(const rust::Vec<BlockStatePair> &pairs,
                                                      endstone::BlockStates *out) noexcept
{
    for (const auto &pair : pairs) {
        switch (pair.value_kind) {
        case kBlockStateBoolean:
            out->emplace(std::string(pair.key), pair.boolean);
            break;
        case kBlockStateText:
            out->emplace(std::string(pair.key), std::string(pair.text));
            break;
        case kBlockStateInteger:
            out->emplace(std::string(pair.key), pair.integer);
            break;
        default:
            return aegilex::kInvalidArgument;
        }
    }
    return aegilex::kOk;
}

void block_states_to_pairs(const endstone::BlockStates &states, rust::Vec<BlockStatePair> *out) noexcept
{
    for (const auto &[key, value] : states) {
        BlockStatePair pair;
        pair.key = rust::String(key);
        std::visit(
            [&pair](const auto &stored) {
                using Stored = std::decay_t<decltype(stored)>;
                if constexpr (std::is_same_v<Stored, bool>) {
                    pair.value_kind = kBlockStateBoolean;
                    pair.boolean = stored;
                }
                else if constexpr (std::is_same_v<Stored, std::string>) {
                    pair.value_kind = kBlockStateText;
                    pair.text = rust::String(stored);
                }
                else {
                    pair.value_kind = kBlockStateInteger;
                    pair.integer = stored;
                }
            },
            value);
        out->push_back(std::move(pair));
    }
}

} // namespace

Block::Block(std::unique_ptr<endstone::Block> block, endstone::Server *server) noexcept
    : impl(std::make_shared<class Block::impl>(std::move(block), server))
{
}

Block::Block(const endstone::Block &block, endstone::Server *server) noexcept
    : impl(std::make_shared<class Block::impl>(block.clone(), server))
{
}

Block::~Block() noexcept = default;

rust::String Block::getType() const
{
    try {
        return rust::String(impl->block->getType());
    }
    catch (...) {
        return rust::String();
    }
}

void Block::setType(const rust::Str type_id, const bool apply_physics) const
{
    try {
        impl->block->setType(std::string(type_id), apply_physics);
    }
    catch (...) {
    }
}

std::int32_t Block::getX() const
{
    try {
        return impl->block->getX();
    }
    catch (...) {
        return 0;
    }
}

std::int32_t Block::getY() const
{
    try {
        return impl->block->getY();
    }
    catch (...) {
        return 0;
    }
}

std::int32_t Block::getZ() const
{
    try {
        return impl->block->getZ();
    }
    catch (...) {
        return 0;
    }
}

Location Block::getLocation() const
{
    try {
        const auto &location = impl->block->getLocation();
        return Location{.x = location.getX(),
                        .y = location.getY(),
                        .z = location.getZ(),
                        .pitch = location.getPitch(),
                        .yaw = location.getYaw(),
                        .dimension = rust::String(location.getDimension().getName())};
    }
    catch (...) {
        return Location{};
    }
}

BlockData Block::getData() const
{
    BlockData data;
    try {
        const auto block_data = impl->block->getData();
        if (block_data == nullptr) {
            return data;
        }
        data.type_id = rust::String(block_data->getType());
        data.runtime_id = block_data->getRuntimeId();
        block_states_to_pairs(block_data->getBlockStates(), &data.states);
    }
    catch (...) {
    }
    return data;
}

BlockSnapshot Block::captureState() const
{
    BlockSnapshot snapshot;
    try {
        auto state = impl->block->captureState();
        if (state == nullptr) {
            return snapshot;
        }
        snapshot.dimension = rust::String(impl->block->getDimension().getName());
        snapshot.x = state->getX();
        snapshot.y = state->getY();
        snapshot.z = state->getZ();
        snapshot.type_id = rust::String(state->getType());
        const auto data = state->getData();
        if (data != nullptr) {
            snapshot.runtime_id = data->getRuntimeId();
            block_states_to_pairs(data->getBlockStates(), &snapshot.states);
        }
    }
    catch (...) {
    }
    return snapshot;
}

std::unique_ptr<Block> Block::getRelative(const std::int32_t dx, const std::int32_t dy, const std::int32_t dz) const
{
    try {
        if (impl == nullptr || impl->block == nullptr) {
            return std::unique_ptr<Block>();
        }
        auto block = impl->block->getRelative(dx, dy, dz);
        return block == nullptr ? std::unique_ptr<Block>() : std::make_unique<Block>(std::move(block), impl->server);
    }
    catch (...) {
        return std::unique_ptr<Block>();
    }
}

std::unique_ptr<Block> Block::clone() const
{
    try {
        if (impl == nullptr || impl->block == nullptr) {
            return std::unique_ptr<Block>();
        }
        auto block = impl->block->clone();
        return block == nullptr ? std::unique_ptr<Block>() : std::make_unique<Block>(std::move(block), impl->server);
    }
    catch (...) {
        return std::unique_ptr<Block>();
    }
}


endstone::Block *Block::native() const noexcept
{
    return impl == nullptr ? nullptr : impl->block.get();
}

std::uint32_t Block::setData(const BlockData &data, const bool apply_physics) const
{
    try {
        if (impl == nullptr || impl->server == nullptr || impl->block == nullptr) {
            return aegilex::kInvalidArgument;
        }
        const std::string type_name(data.type_id);
        if (type_name.empty()) {
            return aegilex::kInvalidArgument;
        }
        endstone::BlockStates state_map;
        if (const auto status = block_states_from_pairs(data.states, &state_map); status != aegilex::kOk) {
            return status;
        }
        const auto block_data = state_map.empty() ? impl->server->createBlockData(type_name)
                                                  : impl->server->createBlockData(type_name, std::move(state_map));
        if (!block_data) {
            return aegilex::kHostError;
        }
        impl->block->setData(*block_data, apply_physics);
        return aegilex::kOk;
    }
    catch (...) {
        return aegilex::kHostError;
    }
}


} // namespace aegilex::native::level

File: rust/crates/aegilex-runtime/src/host/endstone/level/block.rs
//! Core ABI implementation for `native/bindings/endstone/level/block.h`.

use crate::host::endstone::support::*;

impl crate::core_host::imports::HostBlock for PluginStoreState {
    fn block_get_type(&mut self, self_: u32) -> Result<Result<String, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block.block.get-type")?;
            resolve_block(
                self,
                block_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|block| block.getType())
            .map_err(map_core_host_error)
        })())
    }

    fn block_set_type(
        &mut self,
        self_: u32,
        type_id: String,
        apply_physics: Option<bool>,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block.block.set-type")?;
            resolve_block(
                self,
                block_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|block| block.setType(&type_id, apply_physics.unwrap_or(true)))
            .map_err(map_core_host_error)?;
            Ok(())
        })())
    }

    fn block_get_x(&mut self, self_: u32) -> Result<Result<i32, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block.block.get-x")?;
            resolve_block(
                self,
                block_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|block| block.getX())
            .map_err(map_core_host_error)
        })())
    }

    fn block_get_y(&mut self, self_: u32) -> Result<Result<i32, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block.block.get-y")?;
            resolve_block(
                self,
                block_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|block| block.getY())
            .map_err(map_core_host_error)
        })())
    }

    fn block_get_z(&mut self, self_: u32) -> Result<Result<i32, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block.block.get-z")?;
            resolve_block(
                self,
                block_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|block| block.getZ())
            .map_err(map_core_host_error)
        })())
    }

    fn block_get_location(
        &mut self,
        self_: u32,
    ) -> Result<Result<LocationLocation, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block.block.get-location")?;
            resolve_block(
                self,
                block_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|block| wit_location_native(block.getLocation()))
            .map_err(map_core_host_error)
        })())
    }

    fn block_get_data(&mut self, self_: u32) -> Result<Result<u32, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block.block.get-data")?;
            let data = resolve_block(
                self,
                block_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|block| block.getData())
            .map_err(map_core_host_error)?;
            insert_block_data_resource(self, data).map_err(map_core_host_error)
        })())
    }

    fn block_set_data(
        &mut self,
        self_: u32,
        data: u32,
        apply_physics: Option<bool>,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block.block.set-data")?;
            let data =
                copy_native_block_data(block_data_value(self, data).map_err(map_core_host_error)?);
            let block = resolve_block(
                self,
                block_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map_err(map_core_host_error)?;
            native::status_result(block.setData(&data, apply_physics.unwrap_or(true)))
                .map_err(map_core_host_error)?;
            Ok(())
        })())
    }

    fn block_get_relative(
        &mut self,
        self_: u32,
        dx: i32,
        dy: i32,
        dz: i32,
    ) -> Result<Result<u32, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block.block.get-relative")?;
            let block = resolve_block(
                self,
                block_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|block| block.getRelative(dx, dy, dz))
            .map_err(map_core_host_error)?;
            self.insert_block_resource(block)
                .map_err(map_core_host_error)
        })())
    }

    fn block_clone(&mut self, self_: u32) -> Result<Result<u32, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block.block.clone")?;
            let block = resolve_block(
                self,
                block_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|block| block.clone())
            .map_err(map_core_host_error)?;
            self.insert_block_resource(block)
                .map_err(map_core_host_error)
        })())
    }

    fn drop_block(&mut self, handle: u32) -> Result<(), String> {
        self.drop_resource(handle, ResourceKind::Block)
            .map_err(|error| format!("{error:?}"))
    }
}

// ---------------------------------------------------------------------------
// block-data
// ---------------------------------------------------------------------------

impl crate::core_host::imports::HostBlockData for PluginStoreState {
    fn block_data_get_type(
        &mut self,
        self_: u32,
    ) -> Result<Result<String, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block-data.block-data.get-type")?;
            block_data_value(self, self_)
                .map(|data| data.type_id.clone())
                .map_err(map_core_host_error)
        })())
    }

    fn block_data_get_block_states(
        &mut self,
        self_: u32,
    ) -> Result<Result<Vec<BlockDataBlockStatePair>, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block-data.block-data.get-block-states")?;
            block_data_value(self, self_)
                .map(|data| {
                    data.states
                        .iter()
                        .map(wit_block_state_pair_native)
                        .collect()
                })
                .map_err(map_core_host_error)
        })())
    }

    fn block_data_get_runtime_id(
        &mut self,
        self_: u32,
    ) -> Result<Result<u32, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block-data.block-data.get-runtime-id")?;
            block_data_value(self, self_)
                .map(|data| data.runtime_id)
                .map_err(map_core_host_error)
        })())
    }
}

// ---------------------------------------------------------------------------
// block-type
// ---------------------------------------------------------------------------

impl crate::core_host::imports::HostBlockType for PluginStoreState {
    fn has_item_type(&mut self, type_id: String) -> Result<Result<bool, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block-type.has-item-type")?;
            let mut has_item_type = false;
            let level = resolve_level(self.host.clone()).map_err(map_core_host_error)?;
            native::status_result(level.blockTypeHasItem(&type_id, &mut has_item_type))
                .map_err(map_core_host_error)?;
            Ok(has_item_type)
        })())
    }

    fn create_block_data(
        &mut self,
        type_id: String,
    ) -> Result<Result<u32, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block-type.create-block-data")?;
            let mut data = cxx_level::BlockData {
                type_id: String::new(),
                states: Vec::new(),
                runtime_id: 0,
            };
            let level = resolve_level(self.host.clone()).map_err(map_core_host_error)?;
            native::status_result(level.createBlockData(&type_id, &Vec::new(), &mut data))
                .map_err(map_core_host_error)?;
            insert_block_data_resource(self, data).map_err(map_core_host_error)
        })())
    }
}

// ---------------------------------------------------------------------------
// block-state
// ---------------------------------------------------------------------------

impl crate::core_host::imports::HostBlockState for PluginStoreState {
    fn block_state_get_block(&mut self, self_: u32) -> Result<Result<u32, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block-state.block-state.get-block")?;
            let state = block_state_value(self, self_).map_err(map_core_host_error)?;
            let block = resolve_level(self.host.clone())
                .map_err(map_core_host_error)?
                .getBlock(&state.dimension, state.x, state.y, state.z);
            if block.is_null() {
                return Err(map_core_host_error(HostError::from_status(
                    AEGILEX_NOT_FOUND,
                )));
            }
            self.insert_block_resource(block)
                .map_err(map_core_host_error)
        })())
    }

    fn block_state_get_type(
        &mut self,
        self_: u32,
    ) -> Result<Result<String, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block-state.block-state.get-type")?;
            block_state_value(self, self_)
                .map(|state| state.type_id.clone())
                .map_err(map_core_host_error)
        })())
    }

    fn block_state_set_type(
        &mut self,
        self_: u32,
        type_id: String,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block-state.block-state.set-type")?;
            block_state_value_mut(self, self_)
                .map(|state| state.type_id = type_id)
                .map_err(map_core_host_error)?;
            Ok(())
        })())
    }

    fn block_state_get_data(&mut self, self_: u32) -> Result<Result<u32, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block-state.block-state.get-data")?;
            let data = block_state_value(self, self_)
                .map(|state| cxx_level::BlockData {
                    type_id: state.type_id.clone(),
                    states: state
                        .states
                        .iter()
                        .map(|pair| cxx_level::BlockStatePair {
                            key: pair.key.clone(),
                            value_kind: pair.value_kind,
                            boolean: pair.boolean,
                            text: pair.text.clone(),
                            integer: pair.integer,
                        })
                        .collect(),
                    runtime_id: state.runtime_id,
                })
                .map_err(map_core_host_error)?;
            insert_block_data_resource(self, data).map_err(map_core_host_error)
        })())
    }

    fn block_state_set_data(
        &mut self,
        self_: u32,
        data: u32,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block-state.block-state.set-data")?;
            let data =
                copy_native_block_data(block_data_value(self, data).map_err(map_core_host_error)?);
            let state = block_state_value_mut(self, self_).map_err(map_core_host_error)?;
            state.type_id = data.type_id;
            state.states = data.states;
            state.runtime_id = data.runtime_id;
            Ok(())
        })())
    }

    fn block_state_get_x(&mut self, self_: u32) -> Result<Result<i32, TypesHostError>, String> {
        Ok(block_state_value(self, self_)
            .map(|state| state.x)
            .map_err(map_core_host_error))
    }

    fn block_state_get_y(&mut self, self_: u32) -> Result<Result<i32, TypesHostError>, String> {
        Ok(block_state_value(self, self_)
            .map(|state| state.y)
            .map_err(map_core_host_error))
    }

    fn block_state_get_z(&mut self, self_: u32) -> Result<Result<i32, TypesHostError>, String> {
        Ok(block_state_value(self, self_)
            .map(|state| state.z)
            .map_err(map_core_host_error))
    }

    fn block_state_get_location(
        &mut self,
        self_: u32,
    ) -> Result<Result<LocationLocation, TypesHostError>, String> {
        Ok(block_state_value(self, self_)
            .map_err(map_core_host_error)
            .map(|state| LocationLocation {
                dimension: state.dimension.clone(),
                x: state.x as f32,
                y: state.y as f32,
                z: state.z as f32,
                pitch: 0.0,
                yaw: 0.0,
            }))
    }

    fn block_state_update(
        &mut self,
        self_: u32,
        force: Option<bool>,
        apply_physics: Option<bool>,
    ) -> Result<Result<bool, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block-state.block-state.update")?;
            let state = block_state_value(self, self_).map_err(map_core_host_error)?;
            let mut applied = false;
            native::status_result(
                resolve_level(self.host.clone())
                    .map_err(map_core_host_error)?
                    .updateBlockState(
                        state,
                        force.is_some(),
                        force.unwrap_or(false),
                        apply_physics.is_some(),
                        apply_physics.unwrap_or(false),
                        &mut applied,
                    ),
            )
            .map_err(map_core_host_error)?;
            Ok(applied)
        })())
    }

    fn capture(&mut self, block: u32) -> Result<Result<u32, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "block-state.capture")?;
            let snapshot = resolve_block(
                self,
                block_handle(self, block).map_err(map_core_host_error)?,
            )
            .map(|block| block.captureState())
            .map_err(map_core_host_error)?;
            insert_block_state_resource(self, snapshot).map_err(map_core_host_error)
        })())
    }
}
